<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Введение" data-toc-modified-id="Введение-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Введение</a></span><ul class="toc-item"><li><span><a href="#Цели-исследования" data-toc-modified-id="Цели-исследования-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Цели исследования</a></span></li><li><span><a href="#Доступ-к-базе-данных" data-toc-modified-id="Доступ-к-базе-данных-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Доступ к базе данных</a></span></li><li><span><a href="#Первые-строки-таблиц" data-toc-modified-id="Первые-строки-таблиц-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Первые строки таблиц</a></span></li></ul></li><li><span><a href="#Запросы-и-выводы" data-toc-modified-id="Запросы-и-выводы-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Запросы и выводы</a></span><ul class="toc-item"><li><span><a href="#Посчитайте,-сколько-книг-вышло-после-1-января-2000-года" data-toc-modified-id="Посчитайте,-сколько-книг-вышло-после-1-января-2000-года-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Посчитайте, сколько книг вышло после 1 января 2000 года</a></span></li><li><span><a href="#Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку" data-toc-modified-id="Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Для каждой книги посчитайте количество обзоров и среднюю оценку</a></span></li><li><span><a href="#Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры" data-toc-modified-id="Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры</a></span></li><li><span><a href="#Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками" data-toc-modified-id="Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками</a></span></li><li><span><a href="#Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-48-оценок" data-toc-modified-id="Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-48-оценок-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок</a></span></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Общий вывод</a></span></li></ul></li></ul></div>

## Введение

### Цели исследования

**Цель исследования:** Проанализировать базу данных, чтобы сформулировать ценностное предложение для нового продукта

### Доступ к базе данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [3]:
# чтобы выполнить SQL-запрос, используем Pandas
query = '''SELECT * FROM books'''

con=engine.connect()

pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

### Первые строки таблиц

In [4]:
query = '''SELECT * FROM books LIMIT 5'''

con=engine.connect()

pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [5]:
query = '''SELECT * FROM authors LIMIT 5'''

con=engine.connect()

pd.io.sql.read_sql(sql=text(query), con = con)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [6]:
query = '''SELECT * FROM ratings LIMIT 5'''

con=engine.connect()

pd.io.sql.read_sql(sql=text(query), con = con)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [7]:
query = '''SELECT * FROM reviews LIMIT 5'''

con=engine.connect()

pd.io.sql.read_sql(sql=text(query), con = con)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [8]:
query = '''SELECT * FROM publishers LIMIT 5'''

con=engine.connect()

pd.io.sql.read_sql(sql=text(query), con = con)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

## Запросы и выводы

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

### Посчитайте, сколько книг вышло после 1 января 2000 года

In [9]:
query = '''
    SELECT COUNT(book_id)
    FROM books
    WHERE publication_date > '2000-01-01'
    '''

con=engine.connect()

pd.io.sql.read_sql(sql=text(query), con = con)

count
0    819

**Вывод:** После 1 января 2000 года вышла 819 книга

### Для каждой книги посчитайте количество обзоров и среднюю оценку

In [10]:
query = '''
    SELECT title, COUNT(review_id), AVG(rating)
    FROM books as b
    LEFT JOIN reviews as re ON b.book_id = re.book_id
    LEFT JOIN ratings as ra ON b.book_id = ra.book_id
    GROUP BY 1
    ORDER BY 2 DESC, 3 DESC
    '''

con=engine.connect()

pd.io.sql.read_sql(sql=text(query), con = con)

title  count       avg
0                              Twilight (Twilight  #1)   1120  3.662500
1                  The Hobbit  or There and Back Again    528  4.125000
2                               The Catcher in the Rye    516  3.825581
3    Harry Potter and the Prisoner of Azkaban (Harr...    492  4.414634
4    Harry Potter and the Chamber of Secrets (Harry...    480  4.287500
..                                                 ...    ...       ...
994  Disney's Beauty and the Beast (A Little Golden...      0  4.000000
995                               Leonardo's Notebooks      0  4.000000
996                          Essential Tales and Poems      0  4.000000
997    Anne Rice's The Vampire Lestat: A Graphic Novel      0  3.666667
998                            The Natural Way to Draw      0  3.000000

[999 rows x 3 columns]

**Вывод:** Посчитали количество обзоров и среднюю оценку. Эти данные пригодятся нашим коллегам при ценообразовании.

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [11]:
query = '''
    WITH q as (
    SELECT publisher AS Издательство, 
        COUNT(book_id) AS Количество_книг
    FROM books as b
    LEFT JOIN publishers as p ON b.publisher_id = p.publisher_id
    WHERE num_pages > 50
    GROUP BY publisher
    ORDER BY 2 DESC
    LIMIT 1
    )
    SELECT * from q
    '''

con=engine.connect()

pd.io.sql.read_sql(sql=text(query), con = con)

Издательство  Количество_книг
0  Penguin Books               42

**Вывод:** Издательство, которое выпустило наибольшее число книг, не считая брошюр - Penguin Books

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [12]:
query = '''
    WITH q AS (
    SELECT author AS Автор,
        ROUND(AVG(rating), 2) AS Рейтинг
    FROM books as b
    LEFT JOIN authors as a ON a.author_id = b.author_id
    LEFT JOIN ratings as ra ON ra.book_id = b.book_id
    GROUP BY 1
    HAVING COUNT(rating_id) > 50
    ORDER BY 2 DESC
    LIMIT 1
    )
    SELECT * FROM q
    '''

con=engine.connect()

pd.io.sql.read_sql(sql=text(query), con = con)

Автор  Рейтинг
0  J.K. Rowling/Mary GrandPré     4.29

**Вывод:** Автор с самой высокой средней оценкой книг - J.K. Rowling

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок

In [14]:
query = '''
    WITH q AS (
    SELECT username, COUNT(rating_id)
    FROM ratings
    GROUP BY 1
    HAVING COUNT(rating_id) > 48
    ),
    w AS (
    SELECT q.username, COUNT(re.review_id)
    FROM q
    JOIN reviews as re ON re.username = q.username
    GROUP BY 1
    )
    SELECT AVG(count) FROM w
    '''

con=engine.connect()

pd.io.sql.read_sql(sql=text(query), con = con)

avg
0  24.0

**Вывод:** Среднее количество обзоров от пользователей, которые поставили больше 48 оценок - 24 обзора

### Общий вывод




**Мы нашли:**
- Сколько книг из нашей библиотеки вышло в 21 веке
- Самые популярные книги (с наибольшим количеством оценок)
- Самое производительное издательство
- Самого популярного автора
- Самых активных пользователей

**Рекомендация:**
- Теперь мы знаем самые популярные и высокооцененные книги, а также самых популярных авторов. Можно сделать ставку на их продажу, потому что их популярность не зависит от платформы/сервиса.